reconcile/utils/secret_reader.py

```diff
@@ -27,8 +27,107 @@ class SupportsSecret(Protocol):
     q_format: Optional[str]
 
 
+class SupportsVaultSettings(Protocol):
+    """
+    SupportsVaultSettings defines all attributes needed from app-interface-settings
+    to instantiate a SecretReader
+    """
+
+    vault: bool
+
+
+class TypedSecretReader:
+    def read(self, secret: SupportsSecret) -> dict[str, str]:
+        raise NotImplementedError()
+
+    def read_all(self, secret: SupportsSecret) -> dict[str, str]:
```

Comment:
I was thinking about this more and it seems that there were 2 things that resulted in the need to change the original `SecretReader`:

1. `__init__` relied on some unnecessary data structure (`settings`)
2. `read` and `read_all` accepted a `dict` of args that was tightly coupled to data that was returned from queries, instead of accepting the args directly

If those two things never happened, then you never would have had to make this change. So, you've already fixed 1, does it make sense to fix 2 as well?

I can see several cases in our code where we're actually creating a `dict` right now simply to conform to the required parameters of the current `SecretReader`:

* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/openshift_resources_base.py#L439
* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/dashdotdb_base.py#L24
* https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/dyn_traffic_director.py#L512
* https://github.com/app-sre/qontract-reconcile/blob/master/tools/app_interface_reporter.py

For those use cases, would we need to create some new class that conforms to `SupportsSecret` or start directly creating `VaultSecretV1` objects?

What if we just made this really simple:

```python
def read_all(self, path: str, field: str, version: Optional[int] = None, format: Optional[str] = None) -> dict[str, str]
```

This would support any use case where we have a `dict`, `VaultSecretV1`, or neither.

```python
# some_args is a dict like what we return from queries today
secret_reader.read_all(**some_args)
```

And as a result it's still quite easy to use with Pydantic models like:

```python
# vault_secret could be VaultSecretV1 or other pydantic models
secret_reader.read_all(**vault_secret.dict())
```

I might have missed something, so let me know if this makes sense to you or not. I'm not trying to make this more complicated than it needs to be. I'd like to make sure that these patterns that we're introducing are not going to result in similar issues that result in tight coupling and future refactors.

Reply:
i am all for a broader set of read methods. SupportSecret as input parameter is nice because it integrates well with dataclasses generated from our schemas. i also support the idea of read_all with dedicated parameters. the old ones with the dict still hold value while we have not moved away from the old ways.

if we can add all those functions with different parameter types to both SecretReader and TypedSecret reader there would be no difference in the Protocol both would implement and they can be used interchangably.

Reply:
I think it's mostly a trade-off of whether we want a smaller API that works for all things, but the integration requires maybe an additional step (calling `dict()` on pydantic, or even referencing the members directly), or if we want a larger API for the sake of convenience. The additional methods would presumably all need their own tests and could still require future modifications if the way we access data changes again? Versus the method that I proposed would only ever need to change if we fundamentally changed how secrets are accessed.

My preference in order:

1. Try to avoid any care of the data access layer in our `utils` clients (basically what I've shown in the DI examples and `read_all` above)
2. If we're really strongly against above then I'd find some method of "overloading" (Python doesn't really support it, but the methods can have different names) as acceptable, but might cost more in terms of tests, future maintainability, etc?
3. Separate clients based on whether their data and dependencies come from typed (data query classes) or untyped (dict) sources. I'm pretty strongly against this.

Reply:
So I tried to adjust it according to the suggestions:

- SecretReader is now of type TypedSecretReader and can be used interchangeably. 
- TypedSecretReader supports more methods for different inputs: `read(SupportsSecret)`, `read(path, field, format, version)`

Reply:
Went ahead and added some tests already :crossed_fingers:

Reply:
The approach seems reasonable to me. I understand the desire for backwards compatibility and while I don't love the idea of the convenience methods for handling with query data classes, if they're accompanied by more general methods like you've done, that's good enough for me. I'll resolve this thread.